java/src/org/openqa/selenium/grid/distributor/local/LocalDistributor.java

```diff
@@ -753,7 +753,10 @@ private class NewSessionRunnable implements Runnable {
 
     @Override
     public void run() {
-      boolean loop = true;
+      Set<RequestId> inQueue = sessionQueue.getQueueContents().stream()
+        .map(SessionRequestCapability::getRequestId)
+        .collect(Collectors.toSet());
+      boolean loop = !inQueue.isEmpty();
```

Comment:
This is querying the whole queue on each execution, I would prefer to avoid that.

Reply:
I update the code to read the queue only when the `rejectUnsupportedCaps` flag is set, when the flag is not set there is no need to read it.

Note: There is another change in the last push to avoid creating copies of the `ImmutableCapabilities`.